api/utilities.py

```diff
@@ -137,3 +149,164 @@ async def add_parameter_name_and_datetime(request, parameter_name: str | None, d
+    pattern = re.compile(r"^R\d+/\d+(\.\d+)?/\d+(\.\d+)?$")
+    if not pattern.match("/".join(interval)):
+        raise HTTPException(status_code=400, detail=f"Invalid levels repeating-interval: {'/'.join(interval)}")
+
+    amount_of_intervals = int(interval[0][1:])
+    min_height = float(interval[1])
+    increment_value = float(interval[2])
+
+    # Round to 3 decimals to avoid floating point errors
+    return [round(min_height + i * increment_value, 3) for i in range(amount_of_intervals)]
+
+
+def is_float(element: any) -> bool:
+    if element is None:
+        return False
+    try:
+        float(element)
+        return True
+    except ValueError:
+        return False
+
+
+async def get_unique_values_for_metadata(field: str) -> list[str]:
+    request = dstore.GetTSAGRequest(attrs=[field])
+    response = await get_ts_ag_request(request)
+    return [getattr(i.combo, field) for i in response.groups]
+
+
+def filter_observations_for_z(observations, z):
+    if z:
+        z_values = get_z_levels_or_range(z)
+        if isinstance(z_values, tuple):
+            z_min, z_max = z_values[0], z_values[1]
+            observations = [
+                obs
+                for obs in observations
+                if is_float(obs.ts_mdata.level) and z_min <= float(obs.ts_mdata.level) <= z_max
+            ]
+        else:
+            observations = [
+                obs for obs in observations if is_float(obs.ts_mdata.level) and float(obs.ts_mdata.level) in z_values
+            ]
+    return observations
+
+
+def numeric_sort_key(value: str) -> float:
+    """
+    Converts a string to a float for comparison, returns infinity if the string is not convertible.
+    """
+    try:
+        return float(value)
+    except ValueError:
+        return float("inf")
+
+
+def iso_8601_duration_to_seconds_sort_key(duration: str) -> int:
+    try:
+        seconds = iso_8601_duration_to_seconds(duration)
+    except ValueError:
+        seconds = sys.maxsize
+    return seconds
+
+
+def iso_8601_duration_to_seconds(period: str) -> int:
+    try:
+        duration = isodate.parse_duration(period)
+    except ISO8601Error:
+        raise ValueError(f"Invalid ISO 8601 duration: {period}")
+
+    if isinstance(period, isodate.duration.Duration):
+        # Years and months need special handling
+        years_in_seconds = duration.years * 31556926  # Seconds in year
+        months_in_seconds = duration.months * 2629744  # Seconds in month
+        days_in_seconds = duration.tdelta.days * 24 * 60 * 60
+        seconds_in_seconds = duration.tdelta.seconds
+        total_seconds = years_in_seconds + months_in_seconds + days_in_seconds + seconds_in_seconds
+    else:
+        # It's a simple timedelta, so just get the total seconds
+        total_seconds = duration.total_seconds()
+
+    return int(total_seconds)
+
+
+async def get_iso_8601_range(start: str, end: str) -> list[str] | None:
```

Comment:
Found a bug in this. If the requested range is empty ('PT7H/PT7H'), no filtering on period is done.
The issue is that an empty list is then passed to the Datastore filtering, which means no filtering on that parameter.

Maybe best to throw and error directly from this function (like for start > end)?
And add a test for it, to avoid regressions?

Reply:
Fixed in: https://github.com/EURODEO/e-soh/pull/146/commits/8ef434ce941fe37c565e4872ff6da5028d24d912